tests/api_schema.rs

```diff
@@ -0,0 +1,1728 @@
+use apollo_compiler::schema::ExtendedType;
+use apollo_compiler::validation::Valid;
+use apollo_compiler::Schema;
+use apollo_federation::error::FederationError;
+use apollo_federation::Supergraph;
+
+const INACCESSIBLE_V02_HEADER: &str = r#"
+    directive @link(url: String!, as: String, import: [link__Import], for: link__Purpose) repeatable on SCHEMA
+
+    scalar link__Import
+
+    enum link__Purpose {
+      EXECUTION
+      SECURITY
+    }
+
+    directive @inaccessible on FIELD_DEFINITION | OBJECT | INTERFACE | UNION | ARGUMENT_DEFINITION | SCALAR | ENUM | ENUM_VALUE | INPUT_OBJECT | INPUT_FIELD_DEFINITION
+
+    schema
+      @link(url: "https://specs.apollo.dev/link/v0.2")
+      @link(url: "https://specs.apollo.dev/inaccessible/v0.2")
+    {
+      query: Query
+    }
+"#;
+
+fn inaccessible_to_api_schema(input: &str) -> Result<Valid<Schema>, FederationError> {
+    let sdl = format!("{INACCESSIBLE_V02_HEADER}{input}");
+    let graph = Supergraph::new(&sdl)?;
+    graph.to_api_schema()
+}
+
+#[test]
+fn inaccessible_types_with_accessible_references() {
+    let errors = inaccessible_to_api_schema(
+        r#"
+      # Query types can't be inaccessible
+      type Query @inaccessible {
+        someField: String
+      }
+
+      # Inaccessible object type
+      type Object @inaccessible {
+        someField: String
+      }
+
+      # Inaccessible object type can't be referenced by object field in the API
+      # schema
+      type Referencer1 implements Referencer2 {
+        someField: Object!
+      }
+
+      # Inaccessible object type can't be referenced by interface field in the
+      # API schema
+      interface Referencer2 {
+        someField: Object
+      }
+
+      # Inaccessible object type can't be referenced by union member with a
+      # non-inaccessible parent and no non-inaccessible siblings
+      union Referencer3 = Object
+    "#,
+    )
+    .expect_err("should return validation errors");
+
+    insta::assert_display_snapshot!(errors, @r###"
+    The following errors occurred:
+
+      - Type `Query` is @inaccessible but is the query root type, which must be in the API schema.
+
+      - Type `Object` is @inaccessible but is referenced by `Referencer1.someField`, which is in the API schema.
+
+      - Type `Object` is @inaccessible but is referenced by `Referencer2.someField`, which is in the API schema.
+
+      - Type `Referencer3` is in the API schema but all of its members are @inaccessible.
+    "###);
+}
+
+#[test]
+fn removes_inaccessible_object_types() {
+    let api_schema = inaccessible_to_api_schema(
+        r#"
+      extend schema {
+        mutation: Mutation
+        subscription: Subscription
+      }
+
+      # Non-inaccessible object type
```

Comment:
instead of double negation -> should we just update this comment to `accessible`?